### PR TITLE
AADS, on automl, when I launch python train.py, I get the model evaluation with standard metrics

### PR DESCRIPTION
**Pull request by FlorianLeRoyKili**
Image Segmentation TBD later

### CODE REVIEW

File: kiliautoml/models/_base_model.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def train(
         disable_wandb: bool,
         verbose: int,
         **kwargs,
-    ) -> DictTrainingInfos:
+    ) -> Dict[str, Any]:
```
Comment (PierreLeveau) [inline kiliautoml/models/_base_model.py:43]: You should keep `DictTrainingInfos` and define it as `DictTrainingInfos= Dict[str, Any]`

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 2]
```diff
@@ -154,18 +164,71 @@ def tokenize_and_align_labels(examples):
...
             data_collator=data_collator,  # type: ignore
             tokenizer=tokenizer,
             train_dataset=train_dataset,  # type: ignore
+            eval_dataset=eval_dataset,  # type: ignore
+            compute_metrics=compute_metrics,  # type: ignore
         )
-        output = trainer.train()
+
+        trainer.train()
+        train_metrics = trainer.evaluate(trainer.train_dataset)
```
Comment (PierreLeveau) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:197]: put the evaluation part in another private method

**Review by PierreLeveau — CHANGES_REQUESTED** [review]

File: commands/train.py [hunk 3]
```diff
@@ -229,8 +225,10 @@ def main(
...
 
         else:
-            not_implemented_job(job_name, ml_task)
-        training_losses.append([job_name, training_loss])
+            kili_print("not implemented yet")
+        training_losses.append((job_name, model_evaluation))
 
     kili_print()
-    print(tabulate(training_losses, headers=["job_name", "training_loss"]))
+    for job in training_losses:
```
[the remark this answers is not in the document]
Comment (FlorianLeRoyKili) [reply commands/train.py:232]: I put : 
`for evaluation in model_evaluations:
        model_evaluation = pd.DataFrame(evaluation[1]).transpose()
        print(tabulate(model_evaluation, headers=[evaluation[0]] + list(model_evaluation.columns)))`
What do you think ?

File: kiliautoml/models/_ultralytics_object_detection.py [hunk 4]
```diff
@@ -214,21 +227,16 @@ def _yaml_preparation(
...
         assert (
-            n_val_assets > 0
+            n_train_assets > 8
         ), "Validation set must contain at least 2 assets. max_asset should be > 9"
-        assets_splits = {
-            "train": assets[:n_train_assets],
-            "val": assets[n_train_assets : n_train_assets + n_val_assets],
-            "test": assets[n_train_assets + n_val_assets :],
-        }
+        assets_splits = {"train": assets[:n_train_assets], "val": assets[n_train_assets:]}
```
[the remark this answers is not in the document]
Comment (FlorianLeRoyKili) [reply kiliautoml/models/_ultralytics_object_detection.py:239]: Yes it was not used at all.

File: tests/e2e/fixtures/img_class_project_fixture.json [hunk 5]
```diff
@@ -1 +1 @@
...
+[{"consensusTotCoverage": 0, "id": "cl48g1ata000l0no14mir80c0", "inputType": "IMAGE", "jsonInterface": {"jobRendererWidth": 0.2, "jobs": {"CLASSIFICATION_JOB": {"mlTask": "CLASSIFICATION", "content": {"categories": {"T_SHIRT_TOP": {"name": "T-shirt/top"}, "TROUSER": {"name": "Trouser"}, "PULLOVER": {"name": "Pullover"}, "DRESS": {"name": "Dress"}, "COAT": {"name": "Coat"}, "SANDAL": {"name": "Sandal"}, "SHIRT": {"name": "Shirt"}, "SNEAKER": {"name": "Sneaker"}, "BAG": {"name": "Bag"}, "ANKLE_BOOT": {"name": "Ankle boot"}}, "input": "radio"}, "required": 0, "isChild": false, "instruction": "cloth type"}}}, "minConsensusSize": 1, "reviewCoverage": 0, "title": "Fashion MNINST"}]
```
[the remark this answers is not in the document]
Comment (FlorianLeRoyKili) [reply tests/e2e/fixtures/img_class_project_fixture.json:1]: Yes, the other project was on staging.

File: commands/train.py [hunk 6]
```diff
@@ -229,8 +225,10 @@ def main(
...
+        job_evaluation = pd.DataFrame(job[1]).transpose()
```
[the remark this answers is not in the document]
Comment (FlorianLeRoyKili) [reply commands/train.py]: I manage to remove pandas. Actually, even if it was not in the requirements, it was already used in object detection to read the results.csv
I changed it to use csv (even if pandas is a dependency of yolo so if you do object detection, you have to import it anyhow)

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 7]
```diff
@@ -459,6 +477,56 @@ def _predicted_tokens_to_kili_annotations(
...
+            for label in labels
+        ]
+        true_predictions = [
+            [self.label_list[p] for (p, label_id) in zip(prediction, label) if label_id != -100]
+            for prediction, label in zip(predictions, labels)
+        ]
+        all_metrics = metric.compute(predictions=true_predictions, references=true_labels)
+        return all_metrics
+
+    def model_evaulation(self, trainer):
```
Comment (PierreLeveau) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:496]: evaluation

File: kiliautoml/models/_hugging_face_text_classification_model.py [hunk 8]
```diff
@@ -115,11 +124,15 @@ def tokenize_function(examples):
...
+            compute_metrics=self.compute_metrics,  # type: ignore
         )
-        output = trainer.train()
+        trainer.train()
+        model_evaluation = self.model_evaluation(trainer)
+
         kili_print(f"Saving model to {path_model}")
         trainer.save_model(ensure_dir(path_model))
-        return {"training_loss": output.training_loss}
+        return {key: value for key, value in sorted(model_evaluation.items())}
```
Comment (PierreLeveau) [inline kiliautoml/models/_hugging_face_text_classification_model.py:135]: `dict(sorted(model_evaluation.items()))`

**Review by PierreLeveau — CHANGES_REQUESTED** [review]

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 9]
```diff
@@ -89,7 +95,7 @@ def train(
         kili_print(f"Base model: {model_name}")
         path_dataset = os.path.join(PathHF.dataset_dir(model_repository_dir), "data.json")
 
-        label_list = self._kili_assets_to_hf_ner_dataset(
+        self.label_list = self._kili_assets_to_hf_ner_dataset(
```
[the remark this answers is not in the document]
Comment (FlorianLeRoyKili) [reply kiliautoml/models/_hugging_face_named_entity_recognition_model.py:98]: Unfortunately, I don't think that's possible without recoding an entire HuggingFace Trainer (as compute_metrics is used inside trainer.evaluate where I cannot add parameters...).
I can try if you think it is worth it but that will take a couple of hours.

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 10]
```diff
@@ -154,18 +165,38 @@ def tokenize_and_align_labels(examples):
...
+            eval_dataset=eval_dataset,  # type: ignore
+            compute_metrics=compute_metrics,  # type: ignore
         )
-        output = trainer.train()
+        trainer.train()
+        model_evaluation = self.evaluation(trainer)
         kili_print(f"Saving model to {path_model}")
         trainer.save_model(ensure_dir(path_model))
-        return {"training_loss": output.training_loss}
+        return {key: value for key, value in sorted(model_evaluation.items())}
```
Comment (PierreLeveau) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:199]: forgot this one: `dict(sorted(model_evaluation.items()))`

File: kiliautoml/utils/pytorchvision/trainer.py [hunk 11]
```diff
@@ -34,14 +37,15 @@ def train_model_pytorch(
...
-    dataset_sizes = {x: len(dataloaders[x].dataset) for x in ["train", "val"]}
-
     # Decay LR by a factor of 0.1 every 7 epochs
     scheduler = lr_scheduler.StepLR(optimizer, step_size=7, gamma=0.1)
 
     best_model_wts = copy.deepcopy(model.state_dict())
-    best_acc = 0.0
-    best_loss = float("inf")
+    best_val_acc = epoch_train_acc = corresponding_train_acc = 0.0
+    best_val_precision = best_val_recall = best_val_f1 = []
```
Comment (PierreLeveau) [inline kiliautoml/utils/pytorchvision/trainer.py:45]: won't work, this will make `best_val_precision, best_val_recall,  best_val_f1` have the same values since they are pointing towards the same list. Same below

**Review by PierreLeveau — CHANGES_REQUESTED** [review]

File: kiliautoml/models/_pytorchvision_image_classification.py [hunk 12]
```diff
@@ -104,6 +104,13 @@ def train(
         splits["train"], splits["val"] = train_test_split(
             range(len(labels)), test_size=0.2, random_state=42
         )
+        label_train = [labels[i] for i in splits["train"]]
```
Comment (PierreLeveau) [inline kiliautoml/models/_pytorchvision_image_classification.py:107]: note for later: this should be generalized to all the tasks, with a common method (for splitting and checking the label repartition)

Comment (FlorianLeRoyKili) [reply kiliautoml/models/_pytorchvision_image_classification.py:107]: yes. I think you are right. Next week's workshop with Raphaël will help us decide what to choose.

File: kiliautoml/utils/pytorchvision/trainer.py [hunk 13]
```diff
@@ -34,14 +37,23 @@ def train_model_pytorch(
...
     model = model.to(device)
-    dataset_sizes = {x: len(dataloaders[x].dataset) for x in ["train", "val"]}
-
     # Decay LR by a factor of 0.1 every 7 epochs
     scheduler = lr_scheduler.StepLR(optimizer, step_size=7, gamma=0.1)
 
     best_model_wts = copy.deepcopy(model.state_dict())
-    best_acc = 0.0
-    best_loss = float("inf")
+    best_val_acc = 0.0
```
Comment (PierreLeveau) [inline kiliautoml/utils/pytorchvision/trainer.py:44]: Sorry I missed that: that method is now way to big. I suggest a Metric object that would be initialized, updated, then rendered when the method returns.

**Review by PierreLeveau — APPROVED** [review]

File: kiliautoml/utils/pytorchvision/trainer.py (hunk 13 above)
Comment (FlorianLeRoyKili) [reply kiliautoml/utils/pytorchvision/trainer.py:44]: I can put one dictionary for train metrics and one for val metrics?

**Comment by PierreLeveau** [comment]
I'll let @crsegerie have another review before merging

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 14]
```diff
@@ -104,6 +109,10 @@ def train(
                     "tokens": datasets.Sequence(feature=datasets.Value(dtype="string")),  # type: ignore # noqa
                 }
             ),
+        )[
+            "train"
+        ].train_test_split(  # type: ignore
```
Comment (crsegerie) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:114]: thank you

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 15]
```diff
@@ -154,18 +164,38 @@ def tokenize_and_align_labels(examples):
...
+
         data_collator = DataCollatorForTokenClassification(tokenizer)
+
+        def compute_metrics(eval_preds):
+            metric = datasets.load_metric("seqeval")
+            logits, labels = eval_preds
+            predictions = np.argmax(logits, axis=-1)
+            # Remove ignored index (special tokens) and convert to labels
+            true_labels = [
+                [label_list[label_id] for label_id in label if label_id != -100] for label in labels
```
Comment (crsegerie) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:176]: -100 ?

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 16]
```diff
@@ -265,7 +295,10 @@ def _kili_assets_to_hf_ner_dataset(
 
     def _write_asset(self, job_name, labels_to_ids, handler, asset):
         text = self._get_text_from(asset["content"])
-        annotations = asset["labels"][0]["jsonResponse"][job_name]["annotations"]
+        if job_name in asset["labels"][0]["jsonResponse"]:
```
Comment (crsegerie) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:298]: We should create an Asset class, and this kind of logic should be managed by a method of the class. But ok for now

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 17]
```diff
@@ -459,6 +492,40 @@ def _predicted_tokens_to_kili_annotations(
             text, pp_labels, predicted_probas, tokens, null_category, offset_in_text
         )
 
+    def evaluation(self, trainer):
+        train_metrics = trainer.evaluate(trainer.train_dataset)
+        val_metrics = trainer.evaluate()
```
Comment (crsegerie) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:497]: Why there is no argument here?

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py [hunk 18]
```diff
@@ -459,6 +492,40 @@ def _predicted_tokens_to_kili_annotations(
...
+        model_evaluation = {}
+        nb_train_ent = 0
+        nb_val_ent = 0
+        for label in categories_from_job(self.job):
```
Comment (crsegerie) [inline kiliautoml/models/_hugging_face_named_entity_recognition_model.py:501]: Those kinds of function should be a method in the class Job

File: kiliautoml/models/_hugging_face_text_classification_model.py [hunk 19]
```diff
@@ -216,6 +228,81 @@ def _compute_asset_classification(model_framework, tokenizer, model, asset):
...
+            "precision": train_metrics["eval_precision"][-1],  # type: ignore
+            "recall": train_metrics["eval_recall"][-1],  # type: ignore
+            "f1": train_metrics["eval_f1"][-1],  # type: ignore
+        }
+        model_evaluation["val__overall"] = {
+            "loss": val_metrics["eval_loss"],
+            "accuracy": val_metrics["eval_accuracy"],
+            "precision": val_metrics["eval_precision"][-1],  # type: ignore
+            "recall": val_metrics["eval_recall"][-1],  # type: ignore
+            "f1": val_metrics["eval_f1"][-1],  # type: ignore
```
Comment (crsegerie) [inline kiliautoml/models/_hugging_face_text_classification_model.py:302]: I don't understand why sometimes there is a [-1] and sometimes not

File: kiliautoml/models/_ultralytics_object_detection.py [hunk 20]
```diff
@@ -214,21 +231,16 @@ def _yaml_preparation(
     ):
 
         kili_print("Downloading datasets from Kili")
-        train_val_proportions = [0.8, 0.1]
+        train_val_proportions = [0.8, 0.2]
```
Comment (crsegerie) [inline kiliautoml/models/_ultralytics_object_detection.py:234]: good catch

File: kiliautoml/utils/helpers.py [hunk 21]
```diff
@@ -301,3 +309,29 @@ def not_implemented_job(job_name, ml_task):
         "to select one or multiple jobs."
     )
     raise NotImplementedError
+
+
+def print_evaluation(evaluation: Tuple[str, DictTrainingInfosT]):
```
Comment (crsegerie) [inline kiliautoml/utils/helpers.py:314]: separate this tuple into 2 arguments

File: kiliautoml/utils/helpers.py [hunk 22]
```diff
@@ -301,3 +309,29 @@ def not_implemented_job(job_name, ml_task):
...
+    table = []
+    table_int = []
+    for k, values in evaluation[1].items():
+        table_int.append(k)
+        for key in keys:
+            if key in values.keys():
+                table_int.append(round(values[key], 4))
+            else:
+                table_int.append("nan")
+        table.append(table_int)
```
Comment (crsegerie) [inline kiliautoml/utils/helpers.py:335]: why not using pandas ?

File: kiliautoml/utils/pytorchvision/trainer.py [hunk 23]
```diff
@@ -54,48 +57,119 @@ def train_model_pytorch(
...
+        "f1": best_val_metrics["f1"][-1],  # type:ignore
+    }
+    return model, {key: value for key, value in sorted(model_evaluation.items())}
+
+
+def evaluate(running_loss, y_pred, y_true):
+    evaluation = {}
+    evaluation["loss"] = running_loss / len(y_true)
+    evaluation["acc"] = np.sum(y_pred == y_true) / len(y_true)
+    evaluation["precision"] = np.append(
```
Comment (crsegerie) [inline kiliautoml/utils/pytorchvision/trainer.py:155]: why np.append?

File: kiliautoml/utils/pytorchvision/trainer.py [hunk 24]
```diff
@@ -54,48 +57,119 @@ def train_model_pytorch(
...
+        precision_score(
+            y_true, y_pred, average=None, zero_division=0  # type:ignore
+        ),
+        precision_score(
+            y_true, y_pred, average="weighted", zero_division=0  # type:ignore
```
Comment (crsegerie) [inline kiliautoml/utils/pytorchvision/trainer.py:160]: zero_division=np.nan  is better no ?

File: tests/cmd/generate_image_classification_project_mock_data.py [hunk 25]
```diff
@@ -0,0 +1,52 @@
+import json
```
Comment (crsegerie) [inline tests/cmd/generate_image_classification_project_mock_data.py:1]: I have also created a system in my mr to generate mock data...

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py (hunk 15 above)
Comment (FlorianLeRoyKili) [reply kiliautoml/models/_hugging_face_named_entity_recognition_model.py:176]: add comments

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py (hunk 17 above)
Comment (FlorianLeRoyKili) [reply kiliautoml/models/_hugging_face_named_entity_recognition_model.py:497]: add trainer.eval_dataset

File: kiliautoml/models/_hugging_face_named_entity_recognition_model.py (hunk 18 above)
Comment (FlorianLeRoyKili) [reply kiliautoml/models/_hugging_face_named_entity_recognition_model.py:501]: ticket to create a Class Job

File: kiliautoml/models/_hugging_face_text_classification_model.py (hunk 19 above)
Comment (FlorianLeRoyKili) [reply kiliautoml/models/_hugging_face_text_classification_model.py:302]: -1 correspond to the average metric (add comment)

Comment (FlorianLeRoyKili) [reply kiliautoml/models/_hugging_face_text_classification_model.py:302]: add Dataclass

**Review by crsegerie — APPROVED** [review]
